### PR TITLE
ReflectionTypeLoadException exception handling improvement.

### PR DESCRIPTION
**Pull request by IlshatGaripov**
#### Description
A) I ran into the problem recently when loading specific dll with an algorithm to the compute node with lean engine installed. It appeared that one of the algorithms in assembly depended on different dll - that was not installed on the node and this caused ReflectionTypeLoadException to be thrown when call assembly.GetTypes() inside Loader.GetExtendedTypeNames()

More than that assemblyTypes = e.Types collection when exception is thrown may contain null values and then when passed to the next linq query  NullReferenceException will be thrown unexpectedly (from t in assemblyTypes where t.IsClass... ) To protect from this we need to exclude null values

Similiar problem discussion is here:
https://stackoverflow.com/questions/7889228/how-to-prevent-reflectiontypeloadexception-when-calling-assembly-gettypes

B) Additionally, R# suggests that this block of code can be removed because always false:
if (assembly == null)
                {
                    ...
                }
and this is probably true as I could read from documentation both Load and LoadFrom will either return Assembly if they manage to find and load it or will throw one of possible exceptions . 

#### Motivation and Context
Make it less unexpected errors in lean engine, make the error handling easier and more verbose

#### Types of changes

- [ X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [ X] My code follows the code style of this project.
- [ X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. 
- [ ] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: AlgorithmFactory/Loader.cs [hunk 1]
```diff
@@ -241,12 +241,6 @@ private bool TryCreateILAlgorithm(string assemblyPath, out IAlgorithm algorithmI
                     var assemblyBytes = File.ReadAllBytes(assemblyPath);
                     assembly = Assembly.Load(assemblyBytes, debugInformationBytes);
                 }
-                if (assembly == null)
```
Comment (mchandschuh) [inline AlgorithmFactory/Loader.cs]: Gotta love R#! `Assembly.Load` and `Assembly.LoadFrom` guarantee a non-null result or an exception 👍

File: AlgorithmFactory/Loader.cs [hunk 2]
```diff
@@ -315,7 +309,12 @@ public static List<string> GetExtendedTypeNames(Assembly assembly)
                 }
                 catch (ReflectionTypeLoadException e)
                 {
-                    assemblyTypes = e.Types;
+                    assemblyTypes = e.Types.Where(t => t != null).ToArray();
+
+                    foreach (Exception inner in e.LoaderExceptions)
+                    {
+                        Log.Error($"Loader.GetExtendedTypeNames(): {inner.Message}");
```
Comment (mchandschuh) [inline AlgorithmFactory/Loader.cs]: Do we also want to log `e.Message` here as well?

File: AlgorithmFactory/Loader.cs [hunk 3]
```diff
@@ -315,7 +309,12 @@ public static List<string> GetExtendedTypeNames(Assembly assembly)
                 }
                 catch (ReflectionTypeLoadException e)
                 {
-                    assemblyTypes = e.Types;
+                    assemblyTypes = e.Types.Where(t => t != null).ToArray();
```
Comment (mchandschuh) [inline AlgorithmFactory/Loader.cs]: Very interesting! That's certainly unexpected behavior, I wonder why it behaves that way. I suppose the null entries are the ones where the type didn't load. Perhaps a count of the null entries for the log would also be beneficial? I wonder if `e.LoaderExceptions` has an item for null type, i.e, for each type load failure.

Also, thank you for providing extra context via the linked stack overflow. I think it's good practice to include such a reference either in the source code or in the commit message as well for posterity.

File: AlgorithmFactory/Loader.cs (hunk 2 above)
Comment (IlshatGaripov) [reply AlgorithmFactory/Loader.cs]: This message always says: **Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.'**  I will include similiar message to the logs along with count of the null entries .

File: AlgorithmFactory/Loader.cs (hunk 3 above)
Comment (IlshatGaripov) [reply AlgorithmFactory/Loader.cs]: It seems indeed that null values in collection do match the types that could not be loaded. And for each such type there is an exception object in LoaderExceptions  collection. Thanks for a review. I made the minor edits according to comments received. commit with those changes follows  ->

Comment (IlshatGaripov) [reply AlgorithmFactory/Loader.cs]: I use this to count the number of items could not be loaded:
**var countTypesNotLoaded = e.Types.Count(t => t == null);**
Wondering should this be more efficient (faster):
**var countTypesNotLoaded = e.LoaderExceptions.Length;**
if there are exactly as many LoaderExceptions as the types that could not be loaded ?

Comment (IlshatGaripov) [reply AlgorithmFactory/Loader.cs]: https://docs.microsoft.com/en-us/dotnet/api/system.reflection.reflectiontypeloadexception.loaderexceptions?view=netframework-4.8 says about ReflectionTypeLoadException.LoaderExceptions Property:  The null values in the Types array of this instance line up with the exceptions in this array. So I think it's better use LoaderExceptions.Length.

**Pull request merged**